pkg/vaultclient/vaultclient.go

```diff
@@ -0,0 +1,198 @@
+package vaultclient
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/hashicorp/vault/api"
+)
+
+func New(addr, token string) (*VaultClient, error) {
+	client, err := api.NewClient(&api.Config{Address: addr})
+	if err != nil {
+		return nil, err
+	}
+	client.SetToken(token)
+	return &VaultClient{client}, nil
+}
+
+type VaultClient struct {
+	*api.Client
+}
+
+func (v *VaultClient) GetUserFromAliasName(userName string) (*Entity, error) {
+	rawAliases, err := v.Client.Logical().List("identity/entity-alias/id")
+	if err != nil {
+		return nil, fmt.Errorf("failed to lias aliases: %w", err)
+	}
+
+	var aliases aliasListData
+	if err := v.dataInto(rawAliases.Data, &aliases); err != nil {
+		return nil, err
+	}
+
+	var userID string
+	for _, alias := range aliases.KeyInfo {
+		if alias.Name == userName {
+			userID = alias.CanonicalID
+			break
+		}
+	}
+
+	if userID == "" {
+		return nil, &api.ResponseError{StatusCode: http.StatusNotFound, Errors: []string{fmt.Sprintf("no user alias named %s found", userName)}}
+	}
+
+	return v.GetUserByID(userID)
+}
+
+func (v *VaultClient) ListKV(path string) ([]string, error) {
+	var keyResponse keyResponse
+	if err := v.listInto(insertMetadataIntoPath(path), &keyResponse); err != nil {
+		return nil, err
+	}
+	return keyResponse.Keys, nil
+}
+
+func (v *VaultClient) GetKV(path string) (*KVData, error) {
+	var response KVData
+	return &response, v.readInto(insertDataIntoPath(path), &response)
+}
+
+func (v *VaultClient) UpsertKV(path string, data map[string]string) error {
+	_, err := v.Logical().Write(insertDataIntoPath(path), map[string]interface{}{"data": data})
+	return err
+}
+
+func insertMetadataIntoPath(path string) string {
+	pathSplit := strings.Split(path, "/")
+	return strings.Join(append([]string{pathSplit[0], "metadata"}, pathSplit[1:]...), "/")
+}
+
+func insertDataIntoPath(path string) string {
+	pathSplit := strings.Split(path, "/")
+	return strings.Join(append([]string{pathSplit[0], "data"}, pathSplit[1:]...), "/")
+}
```

Comment:
They are both by themselves not sufficient, because `path` might not contain a slash. I've updated it similiar to your first suggestion and added some tests to capture the expectations